File: Programmers/java/LEVEL_3_가장_먼_노드.java
import java.util.*;

class Solution {
    private int bfs(Map<Integer, List<Integer>> map, boolean[] visit, int x) {
        int maxNum = 0;
        int maxCnt = 0;

        Queue<int[]> q = new LinkedList<>(); //현재 노드, 지나온 간선 수
        visit[x] = true;
        q.add(new int[] {x, 0});

        while (!q.isEmpty()) {
            int[] node = q.poll();

            int cur = node[0];
            int cnt = node[1];

            if (maxNum < cnt) {
                maxNum = cnt;
                maxCnt = 1;
            } else {
                maxCnt++;
            }

            List<Integer> list = map.get(cur);

            for (int i = 0; i < list.size(); i++) {
                if (visit[list.get(i)] == false) {
                    q.add(new int[] {list.get(i), cnt + 1});
                    visit[list.get(i)] = true;
                }
            }
        }

        return maxCnt;
    }

    public int solution(int n, int[][] edge) {
        int answer = 0;
        Map<Integer, List<Integer>> map = new HashMap<>();
        boolean[] visit = new boolean[n + 1];

        for (int i = 0; i < edge.length; i++) {
            int x = edge[i][0];
            int y = edge[i][1];

            if (!map.containsKey(x)) {
                map.put(x, new ArrayList<>());
            }
            map.get(x).add(y);

            if (!map.containsKey(y)) {
                map.put(y, new ArrayList<>());
            }
            map.get(y).add(x);
        }

        answer = bfs(map, visit, 1);
        return answer;
    }
}

File: Programmers/cpp/LEVEL_3_가장_먼_노드.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>

using namespace std;

int bfs(unordered_map<int, vector<int>>& um, vector<bool>& visit, int x) {
	int maxNum = 0;
	int maxCnt = 0;

	queue<pair<int, int>> q; //현재 노드, 지나온 간선 수
	visit[x] = true;
	q.push({ x, 0 });

	while (!q.empty()) {
		int cur = q.front().first;
		int cnt = q.front().second;
		q.pop();

		if (maxNum < cnt) {
			maxNum = cnt;
			maxCnt = 1;
		}
		else {
			maxCnt++;
		}

		vector<int> v = um[cur];

		for (int i = 0; i < v.size(); i++) {
			if (visit[v[i]] == false) {
				q.push({ v[i], cnt + 1 });
				visit[v[i]] = true;
			}
		}
	}

	return maxCnt;
}

int solution(int n, vector<vector<int>> edge) {
	int answer = 0;
	unordered_map<int, vector<int>> um;
	vector<bool> visit(n + 1, false);

	for (int i = 0; i < edge.size(); i++) {
		int x = edge[i][0];
		int y = edge[i][1];

		um[x].push_back(y);
		um[y].push_back(x);
	}

	answer = bfs(um, visit, 1);
	return answer;
}